### DISCUSSION TITLE
Combining transformer with diffusion models

### DISCUSSION THREAD

**Discussion by Jack000** [lucidrains/DALLE-pytorch#375] [Show and tell]
Hi guys, just wanted to show you something I've been working on. Here's a link to the code and models: https://github.com/Jack000/guided-diffusion

For a while now I've been thinking about how to combine transformer models with DDPM. The intuition is that diffusion models are great at generating low-level texture but poor at global composition, while transformers have the opposite problem. It would be great if they could be combined somehow.

Anyways after a lot of trial and error I've found an approach that works pretty well. The idea is really simple - just feed the image latents from the VQGAN encoder into a conditional DDPM. So after training a DALLE-pytorch model with the VQGAN VAE, you can feed the image latent code into this diffusion model and get potentially much better image quality.

the diffusion model architecture is also really simple - it's exactly the same as the super-resolution model in guided diffusion, except where they concatenate the low res image channel-wise with the noised input, we can skip the encoder layers entirely and inject the latents into the middle block. This has the added benefit of allowing us to re-use the encoder and decoder weights from the pretrained model, so only the middle block needs to be re-trained.

![arch](https://user-images.githubusercontent.com/2636509/136536649-7e19d687-cb4f-4c05-af34-30980194a103.png)

I also tried some other approaches with varying degrees of success:
- "deblurring" VAE image output with the diffusion model
This was the first and most obvious approach but the images always came out distorted, even with blurring + noise augmentation
- cascaded super resolution
Errors made by the low-resolution model seems to be amplified by the upscaling process. I think you really need the condition augmentation from google SR3 to make this work. Although the 128x128 -> 64x64 -> 256x256 upscaling pipeline seems to work ok
- putting the image embeddings in the encoder layers instead of the middle block
When I tried this the diffusion model quickly converged to produce nearly exactly the same image as the VAE, with errors and all. I think it gets too much information from the skip connections.
- giving CLIP embeddings to the diffusion model as a class embedding
I thought that since most of the original openai models had a class embedding, it would help to give it class information (just replace the nn.embed with a nn.linear) after testing I found that it makes almost zero difference.

So after settling on the current approach, I tried different image embeddings to see how it would affect the diffusion model:
- GumbelVQ f8 from VQGAN (128x128 image size for 256x16x16 latent size)
- a custom trained DVAE from DALLE-pytorch (with same latents as above, 256 embed dimension and 8192 tokens)
- a classifier that uses the DVAE encoder but with the decoder replaced by a classifier head that distills CLIP
I was curious to see what would happen if the latents didn't contain information about reconstruction but semantic content of the image

here are some results from my early experiments:
- ground truth

![init_128](https://user-images.githubusercontent.com/2636509/136545083-6f795f75-7a86-4c5d-b9b3-3ec877f8b79c.png)
a person with glasses is a particularly difficult case so I've been using this image from unsplash for testing

- GumbelVQ f8

![128_progress_00000](https://user-images.githubusercontent.com/2636509/136544878-2d97505e-44ab-43c2-afb4-6f07d68d5096.png) ![128_progress_00001](https://user-images.githubusercontent.com/2636509/136544880-799ae294-c7c2-4de2-9bc3-51114cd0cd01.png) ![128_progress_00002](https://user-images.githubusercontent.com/2636509/136544881-ca71a3a4-09fd-40b4-af0f-aad670071ae6.png) ![128_progress_00003](https://user-images.githubusercontent.com/2636509/136544883-260afd50-83dc-4669-99f1-d740b2449e00.png) ![128_progress_00004](https://user-images.githubusercontent.com/2636509/136544884-a7020422-f326-4f97-bc13-1d0d642dfc30.png) ![128_progress_00005](https://user-images.githubusercontent.com/2636509/136544885-6db856c4-aa91-4a2c-bd53-14f29514638f.png)

- DVAE:

![128_progress_00000](https://user-images.githubusercontent.com/2636509/136543491-8ddd734c-ca35-4494-8d17-5016adc195d6.png) ![128_progress_00001](https://user-images.githubusercontent.com/2636509/136543494-b9a0c8a9-5b31-41a2-8945-cd05553abcf8.png) ![128_progress_00002](https://user-images.githubusercontent.com/2636509/136543495-5e12b45c-f640-46e4-95e8-27840f6ba630.png) ![128_progress_00003](https://user-images.githubusercontent.com/2636509/136543497-74dd2400-ee47-4e0f-b097-278cc3b07231.png) ![128_progress_00004](https://user-images.githubusercontent.com/2636509/136543499-14c7d0f3-9f06-46ec-accb-30b286710e5c.png) ![128_progress_00005](https://user-images.githubusercontent.com/2636509/136543500-26beb44e-0fdf-40b9-8e13-756009fd3745.png)

- Classifer:

![64_progress_00000](https://user-images.githubusercontent.com/2636509/136545598-3e24ca0a-e2ab-4260-b065-6206a1b01246.png) ![64_progress_00001](https://user-images.githubusercontent.com/2636509/136545599-f028df50-0aee-4d5b-aab5-d09fc931e614.png) ![64_progress_00002](https://user-images.githubusercontent.com/2636509/136545601-f662572c-e8f0-44a2-aa4c-8ad52bbaa0bf.png) ![64_progress_00003](https://user-images.githubusercontent.com/2636509/136545603-655d0e4e-277b-40bb-b89c-10539c34daa3.png) ![64_progress_00004](https://user-images.githubusercontent.com/2636509/136545604-2c165698-b8fa-44c1-a26f-8374b9956f03.png) ![64_progress_00005](https://user-images.githubusercontent.com/2636509/136545605-5de140ec-7caf-4d66-999d-dfd94dbe7f1f.png)

I tested on a lot more images, but overall I think the Gumbel f8 embeddings work better for reconstruction. The classifier approach is more like a complete re-interpretation and looks very GAN-like. More training might help but I imagine it would end up similar to regular clip-guided-diffusion.

I also tried training the embeddings together with the diffusion model, from scratch. The codebook collapsed and didn't move much - I think there's a conceptual issue with this, since the diffusion model is training for denoising and we want the embeddings to train for reconstruction, which are not totally aligned goals.

Some generations from the current model:
- 256x256 no clip guidance

![256_progress_00000](https://user-images.githubusercontent.com/2636509/136548567-e04e52a0-8e50-4514-ace4-ddbc699bb809.png) ![256_progress_00001](https://user-images.githubusercontent.com/2636509/136548570-e5e1be57-7bcf-486f-a098-cd863df4f317.png) ![256_progress_00002](https://user-images.githubusercontent.com/2636509/136548572-b2212574-c829-4f3a-becd-e165028735f3.png) ![256_progress_00003](https://user-images.githubusercontent.com/2636509/136548574-2f97a8d9-c516-4a54-b61c-127e649f7632.png) ![256_progress_00004](https://user-images.githubusercontent.com/2636509/136548575-37c0d306-75e2-43e6-a6d0-1a5044986342.png) ![256_progress_00005](https://user-images.githubusercontent.com/2636509/136548576-f5421e42-bc49-4181-9ade-ae5bd2dc46f5.png)

- 256x256 with clip (prompt: a girl wearing glasses)

![256_progress_00000](https://user-images.githubusercontent.com/2636509/136552383-2ae1682e-fde0-42fa-93ca-30883e3321dc.png) ![256_progress_00001](https://user-images.githubusercontent.com/2636509/136552386-f463bcc4-7cb5-4bfa-83a7-30ff752aceb4.png) ![256_progress_00002](https://user-images.githubusercontent.com/2636509/136552389-3b279ab2-c945-415d-8fa7-52f69b3af835.png) ![256_progress_00003](https://user-images.githubusercontent.com/2636509/136552391-fedb12ed-64c4-4f38-b253-96780c5145d9.png) ![256_progress_00004](https://user-images.githubusercontent.com/2636509/136552393-8155ec53-eb7a-45aa-acf3-5a3986763886.png) ![256_progress_00005](https://user-images.githubusercontent.com/2636509/136552394-5e2515a5-94be-4562-987f-92a73c741def.png)

so I find it interesting that the initial noise matters a lot more than clip guidance. I haven't been tracking FID scores or anything so it's very possible that the model is still under-trained.

- 64x64 upscaled to 256x246

![256_progress_00000](https://user-images.githubusercontent.com/2636509/136556248-d8400719-0f62-4f40-bbda-1a68950d2e90.png) ![256_progress_00000](https://user-images.githubusercontent.com/2636509/136556420-8b9fc63f-331b-4c7d-90d8-172c197a7624.png) ![256_progress_00000](https://user-images.githubusercontent.com/2636509/136556610-a4261c4e-38f9-4003-be39-e434da682905.png) ![256_progress_00000](https://user-images.githubusercontent.com/2636509/136556799-37d31fec-b417-475d-8717-1127fbb5a336.png) ![256_progress_00000](https://user-images.githubusercontent.com/2636509/136556974-db87340f-e91a-46ff-a71e-d55ef8830080.png) ![256_progress_00000](https://user-images.githubusercontent.com/2636509/136557188-6ddf39c0-ec48-41ed-9d21-4c8aaf7119ad.png)

- 128x128 downscaled to 64x64 then upscaled to 256x256

![256_progress_00000](https://user-images.githubusercontent.com/2636509/136557609-c8343ad3-7806-4353-b8b8-6035629e5bee.png) ![256_progress_00000](https://user-images.githubusercontent.com/2636509/136557842-94a625e7-6f6c-46d2-82f0-902c83bcf66c.png) ![256_progress_00000](https://user-images.githubusercontent.com/2636509/136558049-228d6682-7765-4fe8-b975-87e66a9f60a9.png) ![256_progress_00000](https://user-images.githubusercontent.com/2636509/136558238-9ca360de-eed4-46d5-aa8e-e52865cdb34e.png) ![256_progress_00000](https://user-images.githubusercontent.com/2636509/136558497-637b3614-dd51-44a1-a17a-3546c4681ab0.png) ![256_progress_00000](https://user-images.githubusercontent.com/2636509/136558701-fe2fe57c-795d-4c05-a1f9-1a3ecb32b2ec.png)

it's a bit of a wash but I think the 256x256 model does a bit better in the mouth and nose areas

There's one experiment that I really wanted to try but couldn't get working - the idea is to train a VQGAN model on an edge-detector version of the image (a Sobel filter would work I think), then feed these edge embeddings to the DDPM. This way the lighting and colors would be entirely generated by the diffusion model, and the transformer would be only responsible for the most salient features of the image. With only edges, the codebook could be a few hundred instead of 8192, possibly enabling 256x24x24 latents which would help a lot with the global structure.

I did try this but couldn't get the VAE to converge (either the DALLE-pytorch DVAE or the VQGAN). There's some kind of issue when the image is predominantly black.

anyways, try it out and let me know what you think. I'd be interested in any suggestions on other vector-quantized image embeddings that we can use. I'm still training the 256x256 model but it's going pretty slowly on 4x3090s

**Comment by iamrajiv** [comment]


**Reply by Jack000** [reply]
ah I just put it on aws s3. I'd recommend cloudflare r2 when that launches, or for now just a regular web server with cloudflare in front of it caching the images

**Reply by Jack000** [reply]
the code should be the same. One tip I'd mention is that I always interleave the images and videos (image, image, video, image, image, video) so that while the user looks at the image the video buffers in the background.

**Reply by Jack000** [reply]
if you already have it online I'd just put cloudflare in front. There are some tutorials online for how to do this with github pages: https://blog.cloudflare.com/secure-and-fast-github-pages-with-cloudflare/

S3 can get expensive if you get a lot of traffic.

**Comment by afiaka87** [comment]
This is a very interesting line of work! I haven't yet grokked the architecture completely, but I'll dig into this when I have some more time. 

Thanks for spending the time; training guided diffusion (in the general case) is indeed an extremely slow endeavor. A fine usage for your 4x3090s!

**Comment by borisdayma** [comment]
This is really cool! Actually I wanted to try it out and asked about it recently (you should join DALLE-Pytorch discord).

I had only thought of the simplest approach: "deblurring" VAE image output with the diffusion model, the VAE being frozen.
I'm a bit disappointed that it didn't work out.

The idea of using the diffusion model early is very interesting but shouldn't you put it a bit after to separate encoding from decoding (so you can have transformers task on encoded tokens).

![image](https://user-images.githubusercontent.com/715491/136670946-b37cdfab-ad2d-4308-9588-4f14b015fc6b.png)

Also I'm not very familiar with diffusion models but do you have to add noise?

For the simple model that would just "deblur" VAE output. I was thinking that the input would be encoded image and output would be the original but didn't think we would also need to add noise (the noise is already know and is the difference between the 2). If that's the case maybe pix2pix, stylegan, etc models are more appropriate?

For the model where you add it at the start of decoder the noise would basically the one from quantized model (codebook - output of encoder).

**Reply by crowsonkb** [reply]
> The idea of using the diffusion model early is very interesting but shouldn't you put it a bit after to separate encoding from decoding (so you can have transformers task on encoded tokens).

The entire diffusion model is the decoder, actually, the encoder is separate.

> Also I'm not very familiar with diffusion models but do you have to add noise?

Yes, diffusion models need an input that is the same size as their output for a noised image (you start with mean 0 std 1 Gaussian noise). Conditional diffusion models like the ones described here use a *second* input for the condition which is usually the same size as the noise input but apparently it works to add it later in the downsampling stages of the model, as done here. The condition for a diffusion upscaler is the low-res image and the condition for a VAE reconstruction error repairer is either the RGB image output by the VAE decoder or the VAE tokens directly. Actually I'm surprised feeding in the VAE tokens didn't work well, it should, I think!

**Reply by Jack000** [reply]
in the super resolution model x is an image (shape of 3x256x256) but in our case x is an embedding (shape of 256x16x16). The diagram is for the diffusion model only, the VAE is trained separately and frozen.

the diagram came from the google paper so the variable names are a bit different (in the openai code x=image, y=class label and low_res=low resolution image [x in the diagram]) The diagram kind of makes it look like a VAE but diffusion models work completely differently basically.

I think you could put the embeddings after the middle block, but I haven't tried it.

**Reply by Jack000** [reply]
I originally wanted to generate 64x64 images with the DALLE-pytorch VAE, then upscale to 256x256 with a fine-tuned diffusion upscaler.

I fine-tuned the diffusion model on the blurred VAE images for a few days, but most of the images came out like this:
![elon_4700](https://user-images.githubusercontent.com/2636509/136680581-c50cd3f9-e2ee-43be-855d-fa7cfe2036e4.png) ![donald_noclip](https://user-images.githubusercontent.com/2636509/136680625-4f055c47-4c5c-42b0-ae8f-803c02860538.png)

It's possible that it would have worked with more training. It didn't seem to be improving so I didn't go further with that approach

worked ok for non-human images though
![mushroom_default](https://user-images.githubusercontent.com/2636509/136680656-4c85013f-7fc0-416e-b1d1-beae46d48da1.png) ![mushroom_macrophoto](https://user-images.githubusercontent.com/2636509/136680658-a18605c9-cb3f-4875-a3ee-7c50a564a245.png) ![mushroom_painting](https://user-images.githubusercontent.com/2636509/136680659-320f775a-c264-4fdd-af21-23633f2c01e4.png)

**Comment by stayforapple** [comment]
Is it applicable to the speech compression ?

**Comment by realrandolph** [comment]
This is excellent
